sources/api/apiclient/src/uri_resolver.rs

```
@@ -0,0 +1,884 @@
+        let key = parts.next().context(S3UriMissingKeySnafu {
+            input_source: input.input.clone(),
+        })?;
+
+        Ok(S3Uri {
+            bucket: bucket.to_string(),
+            key: key.to_string(),
+        })
+    }
+}
+
+#[async_trait]
+impl UriResolver for S3Uri {
+    async fn resolve(&self) -> ResolverResult<String> {
+        use resolver_error::*;
+        let cfg = aws_config::load_defaults(aws_config::BehaviorVersion::latest()).await;
+        let client = aws_sdk_s3::Client::new(&cfg);
+
+        // 1) check the object size using HEAD request
+        let head_resp = client
+            .head_object()
+            .bucket(&self.bucket)
+            .key(&self.key)
+            .send()
+            .await
+            .context(S3HeadSnafu {
+                bucket: self.bucket.clone(),
+                key: self.key.clone(),
+            })?;
+
+        // 2) check if the object size exceeds the limit
+        if let Some(size) = head_resp.content_length {
+            ensure!(
+                (size as u64) < MAX_SIZE_BYTES,
+                resolver_error::S3ObjectTooLargeSnafu {
+                    size: size as u64,
+                    max_size: MAX_SIZE_BYTES,
+                    bucket: self.bucket.clone(),
+                    key: self.key.clone(),
+                }
+            );
+        }
+
+        // 3) GET the object
+        let resp = client
+            .get_object()
+            .bucket(&self.bucket)
+            .key(&self.key)
+            .send()
+            .await
+            .context(S3GetSnafu {
+                bucket: self.bucket.clone(),
+                key: self.key.clone(),
+            })?;
+
+        // 4) COLLECT the body stream
+        let bytes = resp.body.collect().await.context(S3BodySnafu {
+            bucket: self.bucket.clone(),
+            key: self.key.clone(),
+        })?;
+
+        // 5) UTF-8 decode
+        Ok(String::from_utf8_lossy(&bytes.into_bytes()).into_owned())
```

We shouldn't use `from_utf8_lossy` here and should instead return an error if the decoded content is not utf8. We do this above already.

---

Fixed - Used the same logic in the `resolve()` function for `HttpUri`, which now returns an error if the decoded content is not utf8.